### PR TITLE
fix: refactor time-sensitive widget

### PR DESCRIPTION
**Pull request by davidmsibley**
Reworking the time-sensitive-content widget logic. The bug of not reverting back to a basic widget post-templateRetireDate was largely fixed with the change in the partial. It was happening because we were getting into a place where `showTimeSensitiveContent == false` and `showBasicContent == false`

That's been replaced by a check for the presence of a callToAction. The rest of the refactor revolves around using actual `Date`s in the logic, rather than the strings, and breaking decision making blocks out into unit-test-_able_ functions.

`resolveTemplateStatus` smells a little funky. It's basically a bunch of nested ifs, or a backwards switch statement. I think it reads a little clearer like this, but I'd be ok refactoring it as a switch.

----

Contributor License Agreement adherence:

- [x] This Contribution is under the terms of Individual [Contributor License Agreements][] (and also Corporate Contributor License Agreements to the extent applicable) appearing in the [Apereo CLA roster][].

[Apereo CLA roster]: http://licensing.apereo.org/completed-clas
[Contributor License Agreements]: https://www.apereo.org/licensing/agreements

### CODE REVIEW

File: components/portal/widgets/controllers.js [hunk 1]
```diff
@@ -403,152 +403,133 @@ define(['angular'], function(angular) {
...
-        $scope.showTimeSensitiveContent = true;
+    var resolveDate = function(str, defaultEndOfDay) {
+      var result = null;
+      if (str) {
+        var dateStr =
+          $filter('filterForDateWithYear')(str);
+        if (dateStr) {
+          if (dateStr.indexOf('T') < 0) {
+            if (defaultEndOfDay) {
+              dateStr += 'T23:59:59';
```
Comment (ChristianMurphy) [inline components/portal/widgets/controllers.js]: If at all possible use `moment.endOf('day')` to handle this.

File: components/portal/widgets/controllers.js [hunk 2]
```diff
@@ -403,152 +403,133 @@ define(['angular'], function(angular) {
...
+            } else {
+              dateStr += 'T00:00:00';
```
Comment (ChristianMurphy) [inline components/portal/widgets/controllers.js]: If at all possible use `moment.startOf('day')` to handle this.

File: components/portal/widgets/controllers.js [hunk 3]
```diff
@@ -403,152 +403,133 @@ define(['angular'], function(angular) {
...
-      } else if ($filter('filterDifferenceFromDate')(
-        $scope.activePeriodStartDate) > 0) {
-        $scope.templateStatus = 'upcoming';
-      // If filter returns -1, the period has ended
-      } else if ($filter('filterDifferenceFromDate')(
-        $scope.activePeriodEndDate) === -1) {
-        $scope.templateStatus = 'ended';
+      var now = new Date();
+      var templateIndex = indexWithinTemplateRange(now, callsToAction);
+      if (-1 < templateIndex) {
```
Comment (ChristianMurphy) [inline components/portal/widgets/controllers.js]: :art: please put the variable before the constant :bowing_man:

File: components/portal/widgets/controllers.js [hunk 4]
```diff
@@ -403,152 +403,133 @@ define(['angular'], function(angular) {
...
-              ' said it has variable content, but was missing a '
-              + ' start date.');
-            showBasicWidget();
+      angular.forEach(callsToAction, function(callToAction, index) {
+        if (result === -1) { // We haven't found a call to action
+          var startDate =
+            resolveDate(callToAction.activeDateRange.templateLiveDate);
+          var endDate =
+            resolveDate(callToAction.activeDateRange.templateRetireDate, true);
+          if (startDate && startDate < now // Template started
```
Comment (ChristianMurphy) [inline components/portal/widgets/controllers.js]: again moment is amazing: [`moment.isBefore`](https://momentjs.com/docs/#/query/is-between/)

File: components/portal/widgets/controllers.js [hunk 5]
```diff
@@ -403,152 +403,133 @@ define(['angular'], function(angular) {
...
     };
 
-    var showBasicWidget = function() {
-      $scope.showBasicContent = true;
+    var resolveDaysLeft = function(now, actionEnd) {
+      var result = -1;
+      var oneDay = 1000 * 60 * 60 * 24;
+
+      if (now && actionEnd && now < actionEnd) {
+        var difference = actionEnd.getTime() - now.getTime();
```
Comment (ChristianMurphy) [inline components/portal/widgets/controllers.js]: https://momentjs.com/docs/#/displaying/difference/

**Comment by davidmsibley** [comment]
After the checks complete, I'm going to merge this. We can continue with #593 to introduce Moment.

**Pull request merged**